pkg/armrpc/asyncoperation/worker/worker.go

```diff
@@ -251,27 +251,28 @@ func (w *AsyncRequestProcessWorker) runOperation(ctx context.Context, message *q
 
 		logger.Info("Start processing operation.")
 		result, err := asyncCtrl.Run(asyncReqCtx, asyncReq)
+		// Update the result if an error is returned from the controller.
+		// Check that the result is empty to ensure we don't override it, it shouldn't happen.
+		// Controller should always either return non-empty error or non-empty result, but not both.
```

Comment:
I think I have the same comment I did on the previous PR. I'm reflecting back on previous discussions we've had so please tell me if my memory is faulty.

This isn't a comment on the code, but rather the semantics of error handling, and whether our design decisions have aged well. 

We distinguish between the following case:

- Async action completed successfully.
- Async action failed with a known (expected) error. This isn't a bug in Radius.
- Async action failed with an unknown error. This is possibly a bug in Radius.

This matches the error handling in the front-end. 

- Async action completed successfully. -> `HTTP 2XX`
- Async action failed with a known (expected) error. -> `HTTP 4XX`
- Async action failed with an unknown error. -> `HTTP 5XX`

Relevant to this PR - our principle for logging errors is something like:

> error in logs == Radius is misbehaving. 

This is why a _failed_ async result is logged as info. It's an error that's already been handled and classified. 

---

Re: the code This 3-valued logic. 

- Async action completed successfully. -> Empty result + `nil` err
- Async action failed with a known (expected) error. -> Failed result + `nil` err
- Async action failed with an unknown error. -> Empty result + non-`nil` err

If we **wanted** 2-valued logic, we could consolidate two of those cases and simplify all of the code.

Reply:
I'm not trying to belabor the point, if we can let's spend 10 mins on this in the design meeting. Ideally with @youngbupark present.

Reply:
Let's discuss this in the design meeting, but adding my thoughts here to drive the discussion -

1. We don't log handled errors (`Failed result + nil error` scenario) on the server side at all today, this needs to be fixed. The client side errors are out of our control, but it is important to have visibility into them for several reasons - easier debugging, visibility into patterns in client side errors to get insights into areas users are constantly hitting, detecting an outlier, etc.

2. We aren't always handling errors perfectly - [here](https://github.com/radius-project/radius/blob/main/pkg/corerp/backend/controller/createorupdateresource.go#L81) is an example of where this is a known error but controllers returned instead of handling it. So current implementation is brittle. Here is an issue that I had created to track revisiting the implementation: https://github.com/radius-project/radius/issues/7369.

3. "This is why a failed async result is logged as info. It's an error that's already been handled and classified." -- I disagree with this approach to logging. Any error regardless of its type is logged as error. Filtering them separately is done based on the code.

4. [This additional check](https://github.com/radius-project/radius/blob/8754c50c827e912f38bf824d6ae3de709b1ecef3/pkg/armrpc/asyncoperation/worker/worker.go#L255) was added as extra safety net based on the suggestion of @youngbupark [here](https://github.com/radius-project/radius/pull/7370#issuecomment-2018404593), we know this is not expected and the comment calls it out, I do not have strong opinions on keeping it or removing it.


This PR is intended fix #1 above in the most straightforward way possible because it is a pressing issue, has caused enough frustration during debugging that it needs to be fixed asap. The overall design should be revisited improve this code path - I had created an issue here https://github.com/radius-project/radius/issues/7369.

Reply:
Discussed and decided on moving forward with -
* All failures including client side will be logged as errors.
* Skip info log if error is logged since it is redundant. -- updated [here](https://github.com/radius-project/radius/blob/7669a05bba8714f2c4aa3d5f66328b9713cd7203/pkg/armrpc/asyncoperation/worker/worker.go#L273).